pkg/apis/crd/v1alpha1/types.go

```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
+type Source struct {
+	// Namespace is the source namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the source pod.
+	Pod string `json:"pod,omitempty"`
+	// IP is the source IPv4 or IPv6 address. IP as the source is supported
+	// only for live-traffic Traceflow.
+	IP string `json:"ip,omitempty"`
+}
+
+// Destination describes the destination spec of the traceflow.
+type Destination struct {
+	// Namespace is the destination namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the destination pod, exclusive with destination service.
+	Pod string `json:"pod,omitempty"`
+	// Service is the destination service, exclusive with destination pod.
+	Service string `json:"service,omitempty"`
+	// IP is the destination IPv4 or IPv6 address.
+	IP string `json:"ip,omitempty"`
+}
+
+// IPHeader describes spec of an IPv4 header.
+type IPHeader struct {
+	// Protocol is the IP protocol.
+	Protocol int32 `json:"protocol,omitempty" yaml:"protocol,omitempty"`
+	// TTL is the IP TTL.
+	TTL int32 `json:"ttl,omitempty" yaml:"ttl,omitempty"`
+	// Flags is the flags for IP.
+	Flags int32 `json:"flags,omitempty" yaml:"flags,omitempty"`
+}
+
+// IPv6Header describes spec of an IPv6 header.
+type IPv6Header struct {
+	// NextHeader is the IPv6 protocol.
+	NextHeader *int32 `json:"nextHeader,omitempty" yaml:"nextHeader,omitempty"`
+	// HopLimit is the IPv6 Hop Limit.
+	HopLimit int32 `json:"hopLimit,omitempty" yaml:"hopLimit,omitempty"`
+}
+
+// TransportHeader describes spec of a TransportHeader.
+type TransportHeader struct {
+	ICMP *ICMPEchoRequestHeader `json:"icmp,omitempty" yaml:"icmp,omitempty"`
+	UDP  *UDPHeader             `json:"udp,omitempty" yaml:"udp,omitempty"`
+	TCP  *TCPHeader             `json:"tcp,omitempty" yaml:"tcp,omitempty"`
+}
+
+// ICMPEchoRequestHeader describes spec of an ICMP echo request header.
+type ICMPEchoRequestHeader struct {
+	// ID is the ICMPEchoRequestHeader ID.
+	ID int32 `json:"id,omitempty"`
+	// Sequence is the ICMPEchoRequestHeader sequence.
+	Sequence int32 `json:"sequence,omitempty"`
+}
+
+// UDPHeader describes spec of a UDP header.
+type UDPHeader struct {
+	// SrcPort is the source port.
+	SrcPort int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort int32 `json:"dstPort,omitempty"`
+}
+
+// TCPHeader describes spec of a TCP header.
+type TCPHeader struct {
+	// SrcPort is the source port.
+	SrcPort int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort int32 `json:"dstPort,omitempty"`
+	// Flags are flags in the header.
+	Flags int32 `json:"flags,omitempty"`
+}
+
+// Packet includes header info.
+type Packet struct {
+	SrcIP string `json:"srcIP,omitempty"`
+	DstIP string `json:"dstIP,omitempty"`
+	// Length is the IP packet length (includes the IPv4 or IPv6 header length).
+	Length          int32           `json:"length,omitempty"`
+	IPv6Header      *IPv6Header     `json:"ipv6Header,omitempty"`
```

Comment:
I didn't follow the previous discussion too closely, but it it surprising that we have an `IPv6Header` field, but no corresponding field for IPv4.

Reply:
There weren't any discussion about this before. I guess for ipv4 there weren't any important fields other than src or dst ip.

Reply:
In Traceflow `v1beta1.Packet`, there is a `IPHeader` field. But in PacketCapture, we didn't have the implementation yet.  The filter support for TTL/flags didn't seems very important in the first place, so during the previous discussion, we have removed this field.

Reply:
So do we just want to be able to filter based on IP family, or do we care about filtering based on `NextHeader` / `HopLimit`?
This seems a bit inconsistent.
We also don't have a comment on the struct field to explain what it is used for.

Reply:
(Hi , sorry for the absence before, from now on, i will continue to work on this actively to get it merged into v2.2 )

I feel like it would be better to remove `IPv6Header` and added a `IPFamily (string)` field would be enough. The `IPv6Header` struct is not used for filter.

The only problem is for the version upgrade case. Since `v1beta1.Traceflow` is using a similar struct `v1beta1.Packet`, i was hoping we could reuse this in the future. But i guess it won't be much a problem to keep it separated.

Reply:
We need to decide if we want these to be unified in the end or not. If not, then we just have to focus on PacketCapture here, but we may also have to add a unique prefix (`PacketCapture`) to all these structs. If we want them unified, maybe we should have a plan and see if it actually makes sense?

Reply:
I feel an IP protocol field should be useful. If we have that, maybe we can remove ICMP header?

Traceflow needs to generate a packet, which requires more fields about the packet. Maybe ok not to share packet struct between these two.

Reply:
mark this as resloved as a new 'Protocol` field was added. we can continue to disscuss this in the new thread.